File: app/lib/data/models/ai_usage.dart
enum AIUsageAction {
  fluxImageGeneration('flux_image_generation'),
  bunnyUpload('bunny_upload'),
  remotionRender('remotion_render'),
  byokMetadata('byok_metadata');

  const AIUsageAction(this.apiValue);
  final String apiValue;

  static AIUsageAction fromJson(Object? value) => values.firstWhere(
    (candidate) => candidate.apiValue == value,
    orElse: () => throw FormatException('Unknown AI usage action: $value'),
  );
}

class AIUsageScope {
  const AIUsageScope({required this.userId, required this.projectId, this.orgId});
  final String userId;
  final String projectId;
  final String? orgId;

  factory AIUsageScope.fromJson(Map<String, dynamic> json) => AIUsageScope(
    userId: _requiredString(json, 'userId'),
    projectId: _requiredString(json, 'projectId'),
    orgId: _optionalString(json, 'orgId'),
  );
}

class AIQuotaStatus {
  const AIQuotaStatus({
    required this.scope,
    required this.action,
    required this.billingMode,
    required this.allowed,
    required this.unitLimit,
    required this.unitReserved,
    required this.unitConsumed,
    required this.unitRemaining,
    required this.requiredUnits,
    required this.checkedAt,
    this.entitlementId,
    this.reasonCode,
    this.resetAt,
  });
  final AIUsageScope scope;
  final AIUsageAction action;
  final String billingMode;
  final bool allowed;
  final String? entitlementId;
  final String unitLimit;
  final String unitReserved;
  final String unitConsumed;
  final String unitRemaining;
  final String requiredUnits;
  final String? reasonCode;
  final DateTime? resetAt;
  final DateTime checkedAt;

  factory AIQuotaStatus.fromJson(Map<String, dynamic> json) => AIQuotaStatus(
    scope: AIUsageScope.fromJson(_requiredMap(json, 'scope')),
    action: AIUsageAction.fromJson(json['action']),
    billingMode: _requiredString(json, 'billingMode'),
    allowed: json['allowed'] as bool? ?? false,
    entitlementId: _optionalString(json, 'entitlementId'),
    unitLimit: _unit(json, 'unitLimit'),
    unitReserved: _unit(json, 'unitReserved'),
    unitConsumed: _unit(json, 'unitConsumed'),
    unitRemaining: _unit(json, 'unitRemaining'),
    requiredUnits: _unit(json, 'requiredUnits'),
    reasonCode: _optionalString(json, 'reasonCode'),
    resetAt: _optionalDate(json, 'resetAt'),
    checkedAt: _requiredDate(json, 'checkedAt'),
  );
}

class AIUsagePolicyMetadata {
  const AIUsagePolicyMetadata({
    required this.action,
    required this.billingMode,
    required this.estimatedUnits,
    required this.limitBehavior,
    required this.providerFailureBehavior,
  });
  final AIUsageAction action;
  final String billingMode;
  final String estimatedUnits;
  final String limitBehavior;
  final String providerFailureBehavior;

  factory AIUsagePolicyMetadata.fromJson(Map<String, dynamic> json) =>
      AIUsagePolicyMetadata(
        action: AIUsageAction.fromJson(json['action']),
        billingMode: _requiredString(json, 'billingMode'),
        estimatedUnits: _unit(json, 'estimatedUnits'),
        limitBehavior: _requiredString(json, 'limitBehavior'),
        providerFailureBehavior: _requiredString(
          json,
          'providerFailureBehavior',
        ),
      );
}

class AIUsageSummary {
  const AIUsageSummary({required this.projectId, required this.quotas});
  final String projectId;
  final List<AIQuotaStatus> quotas;
  factory AIUsageSummary.fromJson(Map<String, dynamic> json) => AIUsageSummary(
    projectId: _requiredString(json, 'projectId'),
    quotas: _mapList(json, 'quotas', AIQuotaStatus.fromJson),
  );
}

class AIUsagePreflightResponse {
  const AIUsagePreflightResponse({required this.quota, required this.policy});
  final AIQuotaStatus quota;
  final AIUsagePolicyMetadata policy;
  factory AIUsagePreflightResponse.fromJson(Map<String, dynamic> json) =>
      AIUsagePreflightResponse(
        quota: AIQuotaStatus.fromJson(_requiredMap(json, 'quota')),
        policy: AIUsagePolicyMetadata.fromJson(_requiredMap(json, 'policy')),
      );
}

class AIUsageReservation {
  const AIUsageReservation({
    required this.reservationId,
    required this.idempotencyKey,
    required this.scope,
    required this.action,
    required this.billingMode,
    required this.status,
    required this.units,
    required this.createdAt,
    required this.updatedAt,
    required this.expiresAt,
    this.entitlementId,
    this.provider,
    this.model,
    this.jobId,
    this.providerStartedAt,
  });
  final String reservationId;
  final String idempotencyKey;
  final String? entitlementId;
  final AIUsageScope scope;
  final AIUsageAction action;
  final String billingMode;
  final String status;
  final String units;
  final String? provider;
  final String? model;
  final String? jobId;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime expiresAt;
  final DateTime? providerStartedAt;

  factory AIUsageReservation.fromJson(Map<String, dynamic> json) =>
      AIUsageReservation(
        reservationId: _requiredString(json, 'reservationId'),
        idempotencyKey: _requiredString(json, 'idempotencyKey'),
        entitlementId: _optionalString(json, 'entitlementId'),
        scope: AIUsageScope.fromJson(_requiredMap(json, 'scope')),
        action: AIUsageAction.fromJson(json['action']),
        billingMode: _requiredString(json, 'billingMode'),
        status: _requiredString(json, 'status'),
        units: _unit(json, 'units'),
        provider: _optionalString(json, 'provider'),
        model: _optionalString(json, 'model'),
        jobId: _optionalString(json, 'jobId'),
        createdAt: _requiredDate(json, 'createdAt'),
        updatedAt: _requiredDate(json, 'updatedAt'),
        expiresAt: _requiredDate(json, 'expiresAt'),
        providerStartedAt: _optionalDate(json, 'providerStartedAt'),
      );
}

class AIProviderCostMetadata {
  const AIProviderCostMetadata({
    required this.provider,
    required this.providerAction,
    required this.costUnit,
    required this.confidence,
    required this.capturedAt,
    required this.evidence,
    this.model,
    this.providerRequestId,
    this.estimatedCost,
    this.actualCost,
    this.currency,
    this.inputMp,
    this.outputMp,
    this.pricingTableVersion,
  });
  final String provider;
  final String providerAction;
  final String? model;
  final String? providerRequestId;
  final String? estimatedCost;
  final String? actualCost;
  final String costUnit;
  final String? currency;
  final String? inputMp;
  final String? outputMp;
  final String? pricingTableVersion;
  final String confidence;
  final DateTime capturedAt;
  final Map<String, dynamic> evidence;

  factory AIProviderCostMetadata.fromJson(Map<String, dynamic> json) =>
      AIProviderCostMetadata(
        provider: _requiredString(json, 'provider'),
        providerAction: _requiredString(json, 'providerAction'),
        model: _optionalString(json, 'model'),
        providerRequestId: _optionalString(json, 'providerRequestId'),
        estimatedCost: _optionalUnit(json, 'estimatedCost'),
        actualCost: _optionalUnit(json, 'actualCost'),
        costUnit: _requiredString(json, 'costUnit'),
        currency: _optionalString(json, 'currency'),
        inputMp: _optionalUnit(json, 'inputMp'),
        outputMp: _optionalUnit(json, 'outputMp'),
        pricingTableVersion: _optionalString(json, 'pricingTableVersion'),
        confidence: _requiredString(json, 'confidence'),
        capturedAt: _requiredDate(json, 'capturedAt'),
        evidence: _optionalMap(json, 'evidence'),
      );
}

class AIUsageLedgerEntry {
  const AIUsageLedgerEntry({
    required this.entryId,
    required this.idempotencyKey,
    required this.scope,
    required this.action,
    required this.billingMode,
    required this.event,
    required this.units,
    required this.createdAt,
    required this.metadata,
    this.reservationId,
    this.unitDirection,
    this.providerCost,
    this.jobId,
    this.actorId,
    this.reason,
  });
  final String entryId;
  final String idempotencyKey;
  final String? reservationId;
  final AIUsageScope scope;
  final AIUsageAction action;
  final String billingMode;
  final String event;
  final String units;
  final String? unitDirection;
  final AIProviderCostMetadata? providerCost;
  final String? jobId;
  final String? actorId;
  final String? reason;
  final DateTime createdAt;
  final Map<String, dynamic> metadata;

  factory AIUsageLedgerEntry.fromJson(Map<String, dynamic> json) =>
      AIUsageLedgerEntry(
        entryId: _requiredString(json, 'entryId'),
        idempotencyKey: _requiredString(json, 'idempotencyKey'),
        reservationId: _optionalString(json, 'reservationId'),
        scope: AIUsageScope.fromJson(_requiredMap(json, 'scope')),
        action: AIUsageAction.fromJson(json['action']),
        billingMode: _requiredString(json, 'billingMode'),
        event: _requiredString(json, 'event'),
        units: _unit(json, 'units'),
        unitDirection: _optionalString(json, 'unitDirection'),
        providerCost: json['providerCost'] is Map
            ? AIProviderCostMetadata.fromJson(
                Map<String, dynamic>.from(json['providerCost'] as Map),
              )
            : null,
        jobId: _optionalString(json, 'jobId'),
        actorId: _optionalString(json, 'actorId'),
        reason: _optionalString(json, 'reason'),
        createdAt: _requiredDate(json, 'createdAt'),
        metadata: _optionalMap(json, 'metadata'),
      );
}

class AIUsageHistory {
  const AIUsageHistory({required this.projectId, required this.entries});
  final String projectId;
  final List<AIUsageLedgerEntry> entries;
  factory AIUsageHistory.fromJson(Map<String, dynamic> json) => AIUsageHistory(
    projectId: _requiredString(json, 'projectId'),
    entries: _mapList(json, 'entries', AIUsageLedgerEntry.fromJson),
  );
}

class AIUsagePendingReservations {
  const AIUsagePendingReservations({required this.projectId, required this.reservations});
  final String projectId;
  final List<AIUsageReservation> reservations;
  factory AIUsagePendingReservations.fromJson(Map<String, dynamic> json) =>
      AIUsagePendingReservations(
        projectId: _requiredString(json, 'projectId'),
        reservations: _mapList(
          json,
          'reservations',
          AIUsageReservation.fromJson,
        ),
      );
}

class AIUsagePolicyList {
  const AIUsagePolicyList({required this.policies});
  final List<AIUsagePolicyMetadata> policies;
  factory AIUsagePolicyList.fromJson(Map<String, dynamic> json) =>
      AIUsagePolicyList(
        policies: _mapList(json, 'policies', AIUsagePolicyMetadata.fromJson),
      );
}

class AIUsageSnapshot {
  AIUsageSnapshot({
    required this.projectId,
    required this.summary,
    required this.fetchedAt,
    Map<AIUsageAction, AIUsagePreflightResponse> preflights =
        const <AIUsageAction, AIUsagePreflightResponse>{},
  }) : preflights = Map<AIUsageAction, AIUsagePreflightResponse>.unmodifiable(
         preflights,
       );

  final String projectId;
  final AIUsageSummary summary;
  final DateTime fetchedAt;
  final Map<AIUsageAction, AIUsagePreflightResponse> preflights;

  AIUsageSnapshot withPreflight(
    AIUsageAction action,
    AIUsagePreflightResponse preflight, {
    required DateTime fetchedAt,
  }) {
    final quotas = <AIQuotaStatus>[
      for (final quota in summary.quotas)
        if (quota.action != action) quota,
      preflight.quota,
    ];
    return AIUsageSnapshot(
      projectId: projectId,
      summary: AIUsageSummary(projectId: projectId, quotas: quotas),
      fetchedAt: fetchedAt,
      preflights: {...preflights, action: preflight},
    );
  }
}

String _requiredString(Map<String, dynamic> json, String key) {
  final value = _optionalString(json, key);
  if (value == null) throw FormatException('Missing $key');
  return value;
}

String? _optionalString(Map<String, dynamic> json, String key) {
  final value = json[key];
  if (value == null) return null;
  final text = value.toString().trim();
  return text.isEmpty ? null : text;
}

String _unit(Map<String, dynamic> json, String key) =>
    _optionalUnit(json, key) ?? (throw FormatException('Missing $key'));
String? _optionalUnit(Map<String, dynamic> json, String key) =>
    json[key]?.toString();
Map<String, dynamic> _requiredMap(Map<String, dynamic> json, String key) {
  final value = json[key];
  if (value is! Map) throw FormatException('Missing $key');
  return Map<String, dynamic>.from(value);
}

Map<String, dynamic> _optionalMap(Map<String, dynamic> json, String key) {
  final value = json[key];
  return value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};
}

DateTime _requiredDate(Map<String, dynamic> json, String key) =>
    DateTime.parse(_requiredString(json, key));
DateTime? _optionalDate(Map<String, dynamic> json, String key) {
  final value = _optionalString(json, key);
  return value == null ? null : DateTime.parse(value);
}

List<T> _mapList<T>(
  Map<String, dynamic> json,
  String key,
  T Function(Map<String, dynamic>) parse,
) {
  final value = json[key];
  if (value is! List) throw FormatException('Missing $key');
  return List<T>.unmodifiable(
    value.map((entry) => parse(Map<String, dynamic>.from(entry as Map))),
  );
}

File: lab/api/models/ai_usage.py
"""Typed contracts for managed AI usage, quotas, and provider-cost evidence."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Annotated, Any

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


def _to_camel(value: str) -> str:
    head, *tail = value.split("_")
    return head + "".join(part.capitalize() for part in tail)


Identifier = Annotated[str, Field(min_length=1, max_length=128, pattern=r"^\S+$")]
NonNegativeUnits = Annotated[Decimal, Field(ge=0, max_digits=24, decimal_places=8)]
PositiveUnits = Annotated[Decimal, Field(gt=0, max_digits=24, decimal_places=8)]
NonNegativeCost = Annotated[Decimal, Field(ge=0, max_digits=24, decimal_places=8)]


class AIUsageAction(str, Enum):
    FLUX_IMAGE_GENERATION = "flux_image_generation"
    BUNNY_UPLOAD = "bunny_upload"
    REMOTION_RENDER = "remotion_render"
    BYOK_METADATA = "byok_metadata"


class AIUsageBillingMode(str, Enum):
    MANAGED = "managed"
    BYOK = "byok"


class AIEntitlementStatus(str, Enum):
    ACTIVE = "active"
    DISABLED = "disabled"
    EXPIRED = "expired"


class AIUsageReservationStatus(str, Enum):
    RESERVED = "reserved"
    PROVIDER_STARTED = "provider_started"
    CONSUMED = "consumed"
    RELEASED = "released"
    REFUNDED = "refunded"
    EXPIRED = "expired"


class AIUsageLedgerEvent(str, Enum):
    REQUESTED = "requested"
    RESERVED = "reserved"
    PROVIDER_STARTED = "provider_started"
    COMPLETED = "completed"
    FAILED = "failed"
    CONSUMED = "consumed"
    RELEASED = "released"
    REFUNDED = "refunded"
    EXPIRED = "expired"
    ADMIN_ADJUSTMENT = "admin_adjustment"


class AIUsageUnitDirection(str, Enum):
    DEBIT = "debit"
    CREDIT = "credit"


class ProviderCostConfidence(str, Enum):
    EXACT = "exact"
    ESTIMATED = "estimated"
    UNKNOWN = "unknown"


class ProviderCostUnit(str, Enum):
    CURRENCY = "currency"
    PROVIDER_CREDIT = "provider_credit"


class AIQuotaErrorCode(str, Enum):
    EXHAUSTED = "ai_quota_exhausted"
    RATE_LIMITED = "ai_generation_rate_limited"
    ENTITLEMENT_MISSING = "ai_entitlement_missing"
    RESERVATION_CONFLICT = "ai_reservation_conflict"
    SCOPE_INVALID = "ai_usage_scope_invalid"


class AIQuotaErrorKind(str, Enum):
    QUOTA = "quota"
    RATE_LIMIT = "rate_limit"
    ENTITLEMENT = "entitlement"
    CONFLICT = "conflict"


class AIUsageModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=_to_camel,
        extra="forbid",
        populate_by_name=True,
    )


class AIUsageScope(AIUsageModel):
    user_id: Identifier
    project_id: Identifier
    org_id: Identifier | None = None


class ProviderCostMetadata(AIUsageModel):
    provider: Identifier
    provider_action: Identifier
    model: str | None = Field(default=None, min_length=1, max_length=160)
    provider_request_id: str | None = Field(default=None, min_length=1, max_length=256)
    estimated_cost: NonNegativeCost | None = None
    actual_cost: NonNegativeCost | None = None
    cost_unit: ProviderCostUnit = ProviderCostUnit.CURRENCY
    currency: str | None = Field(default=None, pattern=r"^[A-Z]{3}$")
    input_mp: Annotated[Decimal, Field(ge=0, max_digits=18, decimal_places=6)] | None = None
    output_mp: Annotated[Decimal, Field(ge=0, max_digits=18, decimal_places=6)] | None = None
    pricing_table_version: str | None = Field(default=None, min_length=1, max_length=128)
    confidence: ProviderCostConfidence = ProviderCostConfidence.UNKNOWN
    captured_at: datetime
    evidence: dict[str, Any] = Field(default_factory=dict)

    @field_validator("captured_at")
    @classmethod
    def require_aware_captured_at(cls, value: datetime) -> datetime:
        return _require_aware_datetime(value, "captured_at")

    @model_validator(mode="after")
    def validate_cost_evidence(self) -> "ProviderCostMetadata":
        has_cost = self.estimated_cost is not None or self.actual_cost is not None
        if self.cost_unit is ProviderCostUnit.CURRENCY:
            if has_cost and self.currency is None:
                raise ValueError("currency is required when a monetary provider cost is present")
        elif self.currency is not None:
            raise ValueError("provider-credit costs must not claim a currency")
        if self.confidence is ProviderCostConfidence.EXACT and self.actual_cost is None:
            raise ValueError("exact provider cost requires actual_cost")
        if self.confidence is ProviderCostConfidence.ESTIMATED:
            if self.estimated_cost is None:
                raise ValueError("estimated provider cost requires estimated_cost")
            if self.pricing_table_version is None:
                raise ValueError("estimated provider cost requires pricing_table_version")
            if self.actual_cost is not None:
                raise ValueError("estimated provider cost must not claim actual_cost")
        if self.confidence is ProviderCostConfidence.UNKNOWN and has_cost:
            raise ValueError("unknown provider cost must not include cost values")
        if self.actual_cost is not None and self.confidence is not ProviderCostConfidence.EXACT:
            raise ValueError("actual_cost requires exact confidence")
        return self


class AIEntitlement(AIUsageModel):
    entitlement_id: Identifier
    scope: AIUsageScope
    billing_mode: AIUsageBillingMode
    status: AIEntitlementStatus = AIEntitlementStatus.ACTIVE
    actions: list[AIUsageAction] = Field(min_length=1)
    unit_limit: NonNegativeUnits = Decimal("0")
    unit_reserved: NonNegativeUnits = Decimal("0")
    unit_consumed: NonNegativeUnits = Decimal("0")
    valid_from: datetime
    expires_at: datetime | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("valid_from", "expires_at")
    @classmethod
    def require_aware_timestamps(cls, value: datetime | None, info) -> datetime | None:
        if value is None:
            return None
        return _require_aware_datetime(value, info.field_name)

    @model_validator(mode="after")
    def validate_entitlement(self) -> "AIEntitlement":
        if len(set(self.actions)) != len(self.actions):
            raise ValueError("entitlement actions must be unique")
        if self.expires_at is not None and self.expires_at <= self.valid_from:
            raise ValueError("expires_at must be later than valid_from")
        if self.billing_mode is AIUsageBillingMode.BYOK:
            if self.actions != [AIUsageAction.BYOK_METADATA]:
                raise ValueError("BYOK entitlement only supports byok_metadata")
            if any((self.unit_limit, self.unit_reserved, self.unit_consumed)):
                raise ValueError("BYOK entitlement must not consume managed usage units")
        else:
            if AIUsageAction.BYOK_METADATA in self.actions:
                raise ValueError("managed entitlement must not include byok_metadata")
            if self.unit_reserved + self.unit_consumed > self.unit_limit:
                raise ValueError("reserved and consumed units exceed entitlement limit")
        return self

    @property
    def unit_remaining(self) -> Decimal:
        return self.unit_limit - self.unit_reserved - self.unit_consumed


class AIUsageReservation(AIUsageModel):
    reservation_id: Identifier
    idempotency_key: Identifier
    entitlement_id: Identifier | None = None
    scope: AIUsageScope
    action: AIUsageAction
    billing_mode: AIUsageBillingMode
    status: AIUsageReservationStatus = AIUsageReservationStatus.RESERVED
    units: NonNegativeUnits
    provider: Identifier | None = None
    model: str | None = Field(default=None, min_length=1, max_length=160)
    job_id: Identifier | None = None
    created_at: datetime
    updated_at: datetime
    expires_at: datetime
    provider_started_at: datetime | None = None

    @field_validator("created_at", "updated_at", "expires_at", "provider_started_at")
    @classmethod
    def require_aware_timestamps(cls, value: datetime | None, info) -> datetime | None:
        if value is None:
            return None
        return _require_aware_datetime(value, info.field_name)

    @model_validator(mode="after")
    def validate_reservation(self) -> "AIUsageReservation":
        _validate_mode_and_action(self.billing_mode, self.action)
        if self.billing_mode is AIUsageBillingMode.BYOK:
            raise ValueError("BYOK usage does not create managed reservations")
        if self.billing_mode is AIUsageBillingMode.MANAGED and self.units <= 0:
            raise ValueError("managed reservation requires positive units")
        if self.billing_mode is AIUsageBillingMode.MANAGED and self.entitlement_id is None:
            raise ValueError("managed reservation requires entitlement_id")
        if self.updated_at < self.created_at:
            raise ValueError("updated_at must not be earlier than created_at")
        if self.expires_at <= self.created_at:
            raise ValueError("expires_at must be later than created_at")
        if (
            self.status is AIUsageReservationStatus.RESERVED
            and self.provider_started_at is not None
        ):
            raise ValueError("reserved usage must not have provider_started_at")
        if (
            self.status is not AIUsageReservationStatus.RESERVED
            and self.provider_started_at is None
        ):
            if self.status in {
                AIUsageReservationStatus.PROVIDER_STARTED,
                AIUsageReservationStatus.CONSUMED,
                AIUsageReservationStatus.REFUNDED,
            }:
                raise ValueError(f"{self.status.value} reservation requires provider_started_at")
        return self


class AIUsageLedgerEntry(AIUsageModel):
    entry_id: Identifier
    idempotency_key: Identifier
    reservation_id: Identifier | None = None
    scope: AIUsageScope
    action: AIUsageAction
    billing_mode: AIUsageBillingMode
    event: AIUsageLedgerEvent
    units: NonNegativeUnits = Decimal("0")
    unit_direction: AIUsageUnitDirection | None = None
    provider_cost: ProviderCostMetadata | None = None
    job_id: Identifier | None = None
    actor_id: Identifier | None = None
    reason: str | None = Field(default=None, min_length=1, max_length=500)
    created_at: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("created_at")
    @classmethod
    def require_aware_created_at(cls, value: datetime) -> datetime:
        return _require_aware_datetime(value, "created_at")

    @model_validator(mode="after")
    def validate_ledger_entry(self) -> "AIUsageLedgerEntry":
        _validate_mode_and_action(self.billing_mode, self.action)
        if self.billing_mode is AIUsageBillingMode.BYOK and self.units != 0:
            raise ValueError("BYOK usage must not consume managed usage units")
        reservation_events = {
            AIUsageLedgerEvent.RESERVED,
            AIUsageLedgerEvent.PROVIDER_STARTED,
            AIUsageLedgerEvent.COMPLETED,
            AIUsageLedgerEvent.FAILED,
            AIUsageLedgerEvent.CONSUMED,
            AIUsageLedgerEvent.RELEASED,
            AIUsageLedgerEvent.REFUNDED,
            AIUsageLedgerEvent.EXPIRED,
        }
        unit_events = {
            AIUsageLedgerEvent.RESERVED,
            AIUsageLedgerEvent.CONSUMED,
            AIUsageLedgerEvent.RELEASED,
            AIUsageLedgerEvent.REFUNDED,
            AIUsageLedgerEvent.EXPIRED,
            AIUsageLedgerEvent.ADMIN_ADJUSTMENT,
        }
        if (
            self.billing_mode is AIUsageBillingMode.MANAGED
            and self.event in reservation_events
            and self.reservation_id is None
        ):
            raise ValueError(f"{self.event.value} event requires reservation_id")
        if self.billing_mode is AIUsageBillingMode.BYOK and self.reservation_id is not None:
            raise ValueError("BYOK ledger metadata must not reference a managed reservation")
        if self.event in unit_events and self.units <= 0:
            raise ValueError(f"{self.event.value} event requires positive units")
        if self.event is AIUsageLedgerEvent.ADMIN_ADJUSTMENT:
            if self.actor_id is None or self.reason is None or self.unit_direction is None:
                raise ValueError("admin adjustment requires actor_id, reason, and unit_direction")
        elif self.unit_direction is not None:
            raise ValueError("unit_direction is reserved for admin adjustments")
        if self.provider_cost is not None and self.event not in {
            AIUsageLedgerEvent.COMPLETED,
            AIUsageLedgerEvent.FAILED,
        }:
            raise ValueError("provider cost evidence belongs on completed or failed events")
        return self


class AIQuotaStatus(AIUsageModel):
    scope: AIUsageScope
    action: AIUsageAction
    billing_mode: AIUsageBillingMode
    allowed: bool
    entitlement_id: Identifier | None = None
    unit_limit: NonNegativeUnits = Decimal("0")
    unit_reserved: NonNegativeUnits = Decimal("0")
    unit_consumed: NonNegativeUnits = Decimal("0")
    unit_remaining: NonNegativeUnits = Decimal("0")
    required_units: NonNegativeUnits = Decimal("0")
    reason_code: AIQuotaErrorCode | None = None
    reset_at: datetime | None = None
    checked_at: datetime

    @field_validator("reset_at", "checked_at")
    @classmethod
    def require_aware_timestamps(cls, value: datetime | None, info) -> datetime | None:
        if value is None:
            return None
        return _require_aware_datetime(value, info.field_name)

    @model_validator(mode="after")
    def validate_quota_status(self) -> "AIQuotaStatus":
        _validate_mode_and_action(self.billing_mode, self.action)
        if self.billing_mode is AIUsageBillingMode.BYOK:
            if any(
                (
                    self.unit_limit,
                    self.unit_reserved,
                    self.unit_consumed,
                    self.unit_remaining,
                    self.required_units,
                )
            ):
                raise ValueError("BYOK quota status must not expose managed usage units")
        else:
            calculated_remaining = self.unit_limit - self.unit_reserved - self.unit_consumed
            if calculated_remaining < 0 or self.unit_remaining != calculated_remaining:
                raise ValueError("unit_remaining must match the non-negative entitlement balance")
            if self.allowed and self.entitlement_id is None:
                raise ValueError("allowed managed usage requires entitlement_id")
            if self.allowed and self.required_units > self.unit_remaining:
                raise ValueError("allowed usage cannot require more units than remain")
        if not self.allowed and self.reason_code is None:
            raise ValueError("blocked quota status requires reason_code")
        if self.allowed and self.reason_code is not None:
            raise ValueError("allowed quota status must not include reason_code")
        return self


class AIQuotaError(AIUsageModel):
    code: AIQuotaErrorCode
    kind: AIQuotaErrorKind
    message: str = Field(min_length=1, max_length=500)
    action: AIUsageAction
    billing_mode: AIUsageBillingMode
    provider: Identifier | None = None
    remaining_units: NonNegativeUnits | None = None
    required_units: PositiveUnits | None = None
    retryable: bool
    retry_after_seconds: int | None = Field(default=None, gt=0)
    settings_path: str | None = Field(default=None, min_length=1, max_length=256)
    details: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_error_envelope(self) -> "AIQuotaError":
        _validate_mode_and_action(self.billing_mode, self.action)
        expected_kind = {
            AIQuotaErrorCode.EXHAUSTED: AIQuotaErrorKind.QUOTA,
            AIQuotaErrorCode.RATE_LIMITED: AIQuotaErrorKind.RATE_LIMIT,
            AIQuotaErrorCode.ENTITLEMENT_MISSING: AIQuotaErrorKind.ENTITLEMENT,
            AIQuotaErrorCode.RESERVATION_CONFLICT: AIQuotaErrorKind.CONFLICT,
            AIQuotaErrorCode.SCOPE_INVALID: AIQuotaErrorKind.ENTITLEMENT,
        }[self.code]
        if self.kind is not expected_kind:
            raise ValueError(f"{self.code.value} requires kind {expected_kind.value}")
        if self.code is AIQuotaErrorCode.EXHAUSTED:
            if self.remaining_units is None or self.required_units is None:
                raise ValueError("quota exhaustion requires remaining_units and required_units")
            if self.required_units <= self.remaining_units:
                raise ValueError("quota exhaustion requires required_units above remaining_units")
            if self.retryable:
                raise ValueError("quota exhaustion is not retryable without an entitlement change")
        if self.code is AIQuotaErrorCode.RATE_LIMITED:
            if not self.retryable or self.retry_after_seconds is None:
                raise ValueError("rate limit errors require retryable and retry_after_seconds")
        elif self.retry_after_seconds is not None:
            raise ValueError("retry_after_seconds is reserved for rate limit errors")
        if self.code is AIQuotaErrorCode.ENTITLEMENT_MISSING and self.retryable:
            raise ValueError("missing entitlement is not retryable without user action")
        return self


_ALLOWED_RESERVATION_TRANSITIONS: dict[
    AIUsageReservationStatus,
    frozenset[AIUsageReservationStatus],
] = {
    AIUsageReservationStatus.RESERVED: frozenset(
        {
            AIUsageReservationStatus.PROVIDER_STARTED,
            AIUsageReservationStatus.RELEASED,
            AIUsageReservationStatus.EXPIRED,
        }
    ),
    AIUsageReservationStatus.PROVIDER_STARTED: frozenset(
        {
            AIUsageReservationStatus.CONSUMED,
            AIUsageReservationStatus.RELEASED,
            AIUsageReservationStatus.REFUNDED,
        }
    ),
    AIUsageReservationStatus.CONSUMED: frozenset({AIUsageReservationStatus.REFUNDED}),
    AIUsageReservationStatus.RELEASED: frozenset(),
    AIUsageReservationStatus.REFUNDED: frozenset(),
    AIUsageReservationStatus.EXPIRED: frozenset(),
}


def validate_reservation_transition(
    current: AIUsageReservationStatus,
    target: AIUsageReservationStatus,
) -> None:
    if target not in _ALLOWED_RESERVATION_TRANSITIONS[current]:
        raise ValueError(f"invalid reservation transition: {current.value} -> {target.value}")


def _validate_mode_and_action(
    billing_mode: AIUsageBillingMode,
    action: AIUsageAction,
) -> None:
    if billing_mode is AIUsageBillingMode.BYOK:
        if action is not AIUsageAction.BYOK_METADATA:
            raise ValueError("BYOK mode only supports byok_metadata")
        return
    if action is AIUsageAction.BYOK_METADATA:
        raise ValueError("managed mode must not use byok_metadata")


def _require_aware_datetime(value: datetime, field_name: str) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{field_name} must include a timezone")
    return value
